cdc/sink/codec/canal.go

```diff
@@ -0,0 +1,362 @@
+		RowDatas:         nil,
+		DdlSchemaName:    e.Schema,
+	}
+	rcBytes, err := proto.Marshal(rc)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// build entry
+	entry := &canal.Entry{
+		Header:           header,
+		EntryTypePresent: &canal.Entry_EntryType{EntryType: canal.EntryType_ROWDATA},
+		StoreValue:       rcBytes,
+	}
+	return entry, nil
+}
+
+// NewCanalEntryBuilder creates a new canalEntryBuilder
+func NewCanalEntryBuilder() *canalEntryBuilder {
+	d := charmap.ISO8859_1.NewDecoder()
+	return &canalEntryBuilder{
+		bytesDecoder: d,
+	}
+}
+
+// CanalEventBatchEncoder encodes the events into the byte of a batch into.
+type CanalEventBatchEncoder struct {
+	messages     *canal.Messages
+	packet       *canal.Packet
+	entryBuilder *canalEntryBuilder
+}
+
+// AppendResolvedEvent implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) AppendResolvedEvent(ts uint64) error {
+	// For canal now, there is no such a corresponding type to ResolvedEvent so far.
+	// Therefore the event is ignored.
+	return nil
+}
+
+// AppendRowChangedEvent implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) AppendRowChangedEvent(e *model.RowChangedEvent) error {
+	entry, err := d.entryBuilder.FromRowEvent(e)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	b, err := proto.Marshal(entry)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	d.messages.Messages = append(d.messages.Messages, b)
+	return nil
+}
+
+// AppendDDLEvent implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) AppendDDLEvent(e *model.DDLEvent) error {
+	entry, err := d.entryBuilder.FromDdlEvent(e)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	b, err := proto.Marshal(entry)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	d.messages.Messages = append(d.messages.Messages, b)
+	return nil
+}
+
+// Build implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) Build() (key []byte, value []byte) {
+	err := d.refreshPacketBody()
+	if err != nil {
+		panic(err)
+	}
+	value, err = proto.Marshal(d.packet)
+	if err != nil {
+		panic(err)
+	}
+	return nil, value
+}
+
+// Size implements the EventBatchEncoder interface
+func (d *CanalEventBatchEncoder) Size() int {
+	// TODO: avoid marshaling the messages every time for calculating the size of the packet
+	err := d.refreshPacketBody()
+	if err != nil {
+		panic(err)
+	}
+	return proto.Size(d.packet)
+}
+
+// refreshPacketBody() marshals the messages to the packet body
+func (d *CanalEventBatchEncoder) refreshPacketBody() error {
+	oldSize := len(d.packet.Body)
+	newSize := proto.Size(d.messages)
+	if newSize > oldSize {
+		// resize packet body slice
+		d.packet.Body = append(d.packet.Body, make([]byte, newSize-oldSize)...)
+	}
+	_, err := d.messages.MarshalToSizedBuffer(d.packet.Body[:newSize])
+	return err
+}
+
+// NewCanalEventBatchEncoder creates a new CanalEventBatchEncoder.
+func NewCanalEventBatchEncoder() EventBatchEncoder {
```

Comment:
could you add a unit test for canal event batch encoder?

Reply:
unit tests added in https://github.com/pingcap/ticdc/pull/649/commits/c6079f5a8826846fd9c6330eda302b80b432c905